resource-aggregate/events/deviceMetadataSnapshotTaken.go

```diff
@@ -84,11 +84,7 @@ func (d *DeviceMetadataSnapshotTaken) HandleDeviceMetadataUpdated(ctx context.Co
 	}
 
 	d.DeviceId = upd.GetDeviceId()
-	if d.DeviceMetadataUpdated == nil {
-		d.DeviceMetadataUpdated = upd
-	} else {
-		d.DeviceMetadataUpdated.CopyData(upd)
-	}
+	d.DeviceMetadataUpdated = upd
```

Comment:
We still have a bit of dangerous code relating to the `DeviceMetadataUpdated` struct:

```
func (p *deviceMetadataProjection) GetDeviceMetadataUpdated() *events.DeviceMetadataUpdated {
	p.private.lock.RLock()
	defer p.private.lock.RUnlock()
	return p.private.snapshot.GetDeviceMetadataUpdated()
}
```

This only locks the snapshot and it is still possible to get the same `DeviceMetadataUpdated` pointer in multiple threads. Only if the pointers are used as read-only (eg the member variables are never modified) it is safe, but there is no `const` in golang, so we cannot guarantee it. To make it 100% safe we would have to return a deep copy from `GetDeviceMetadataUpdated() `.

Reply:
`p.private.snapshot.DeviceMetadataUpdated` must be immutable and this part of code is generated by protobuf.